CASTOR reconstruction and simulation updates

hvanhaev opened the pull request:
This pull request contains updates for CASTOR reconstruction and simulation.
After integration of this in master, the request is that these changes are also allowed to be back ported to the 76X and 80X branches for use with 2015 and 2016 data.

Reconstruction:
* Fix of RecHitCorrector class: 
   - this class is not used during default reconstruction sequence, but it is important for analyses, as it can very easily reprocess CASTOR objects on the fly during ntuple steps. 
   - the code is updated with: memory leak fix, removal of obsolete factor 2, cleaning up code following comments from a previous PR (https://github.com/cms-sw/cmssw/pull/21860#issuecomment-357963218). 
   - I tried to replace the for loops with range-based for loop, but could not figure it out for the line "for (unsigned int i=0;i<rechits->size();i++)". If someone can provide an alternative for this then please let me know.

Simulation:
* A long time ago we discovered a small bug in the CASTOR noise addition (during digi step), which would overestimate the noise, especially in case of pA and AA collisions. Now finally I managed to prepare the code for submission. This change has already been validated within CASTOR group. These classes are used during production whenever the CASTOR simulation is executed/included in the config.

hvanhaev committed:
clean up + remove obsolete factor 2 + fix memory leak

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22424/3679

cmsbuild commented:
A new Pull Request was created by  @hvanhaev (Hans Van Haevermaet) for master.

It involves the following packages:

RecoLocalCalo/Castor


@perrotta, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

hvanhaev committed:
CASTOR noise fix in digitisation

hvanhaev committed:
Merge branch 'from-CMSSW_10_1_0_pre1' of https://github.com/hvanhaev/cmssw into from-CMSSW_10_1_0_pre1

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22424/3833

cmsbuild commented:
Pull request #22424 was updated. @perrotta, @cmsbuild, @civanch, @mdhildreth, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

hvanhaev commented:
Ok, so I unintentionally added a commit to this pull request on a different topic, didn't know GitHub did this automatically. Anyway, it also needs to find its way into CMSSW, but it is related to simulation group: a long time ago we discovered a small bug in the CASTOR noise addition (during digi step), which would overestimate the noise, especially in case of pA and AA collisions. Now finally I managed to prepare the code for submission. This change has already been validated within CASTOR group. Also here, we eventually like to be able to back port this into 76X and 80X branches. Please let me know if I need to separate the topics(commits) or not. Apologies if I caused any confusion.

perrotta commented:
github don't do it automatically, but only if you push the change into the branch ;-)

In principle, approval and integration can proceed faster if different and unrelated issues are kept separated, and this is what I'd suggest here.

Anyhow, if both reco and sim updates can be considered complete here (modulo the review) perhaps you can also stick on this single pull request. If you do so, however, please update the title and add a complete description of the two by editing the description window on top of this thread

By the way, could you please also specify in the description what is the expected usage for the RecHitCorrector class and all other classes in ```RecoLocalCalo/Castor/src```? 
If they are supposed to be used in production, please convert them into thread safe stream modules (as in was already suggested in a previous incarnation of this PR in https://github.com/cms-sw/cmssw/pull/21860#issuecomment-357963218); and moreover, please move the class declarations in the ```\interface``` subdirectory, and only keep the implementation in ```\src``` (this suggestion holds also if these classes are not meant to be used in any central production)

cmsbuild commented:
+1
Tested at: af2458487e0e26a2bbb6eb8ac0f0c30c00993b6f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/26693/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/af2458487e0e26a2bbb6eb8ac0f0c30c00993b6f
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/26693/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/26693/git-merge-result

cmsbuild commented:
Comparison job queued.

hvanhaev commented:
Hi,

I changed the title and updated the description at the top.
Regarding the rest of RecoLocalCalo/Castor: I wanted to avoid to change everything, if I need to clean up the other files it will take me more time. In particular CastorTowerProducer is important yes, since it is executed during default production. The others are not. Concerning your previous comments, I implemented those on the RecHitCorrector, as the PR was about that. Not sure what I still have to convert in the other files to be thread safe there, after quickly looking it only involves removing commented lines it seems.

I will have a look at the other files in /src, once ready I will include it here or in a new PR.

On 8 Mar 2018, at 11:46, perrotta <notifications@github.com<mailto:notifications@github.com>> wrote:


github don't do it automatically, but only if you push the change into the branch ;-)

In principle, approval and integration can proceed faster if different and unrelated issues are kept separated, and this is what I'd suggest here.

Anyhow, if both reco and sim updates can be considered complete here (modulo the review) perhaps you can also stick on this single pull request. If you do so, however, please update the title and add a complete description of the two by editing the description window on top of this thread

By the way, could you please also specify in the description what is the expected usage for the RecHitCorrector class and all other classes in RecoLocalCalo/Castor/src?
If they are supposed to be used in production, please convert them into thread safe stream modules (as in was already suggested in a previous incarnation of this PR in #21860 (comment)<https://github.com/cms-sw/cmssw/pull/21860#issuecomment-357963218>); and moreover, please move the class declarations in the \interface subdirectory, and only keep the implementation in \src (this suggestion holds also if these classes are not meant to be used in any central production)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/22424#issuecomment-371451628>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEgkF6QIAWSQIPn0tnwgFgyoNQvyuLyzks5tcQwfgaJpZM4SaFZH>.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/26693/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 76 differences found in the comparisons
 - DQMHistoTests: Total files compared: 29
 - DQMHistoTests: Total histograms compared: 2480211
 - DQMHistoTests: Total failures: 84
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2479951
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.15000000007 KiB( 23 files compared)
 - Checked 118 log files, 9 edm output root files, 29 DQM output files

perrotta commented:
> I changed the title and updated the description at the top.

Thank you

> Regarding the rest of RecoLocalCalo/Castor: I wanted to avoid to change everything, if I need to clean
> up the other files it will take me more time. In particular CastorTowerProducer is important yes, since it
> is executed during default production. The others are not. Concerning your previous comments, I
> implemented those on the RecHitCorrector, as the PR was about that. Not sure what I still have to
> convert in the other files to be thread safe there, after quickly looking it only involves removing
>  commented lines it seems.

Ok, fine: CastorTowerProducer is already a stream module. I agree with you about not touching now the other modules. Please also disregard my previous suggestion of moving the class declaration into header files in a ```\interface``` subdirectory: since they are producers they don't need to be exposed, and it is correct that they don't have an interface that can be included.

However, since you are touching here RecHitCorrector, please make it a stream module also: it is true that it is not used now in production, but since it stay in a \src area let it move away from being a legacy module (otherwise: move it into a \test area)

perrotta commented on RecoLocalCalo/Castor/src/RecHitCorrector.cc:
```diff
@@ -114,59 +114,47 @@ RecHitCorrector::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
-	//std::cout << "rechit energy(fC) = " << fC << " time = " << time << std::endl;
 	
 	// do proper gain calibration reading the latest entries in the condDB
 	const CastorCalibrations& calibrations=conditions->getCastorCalibrations(rechit.id());
```
```calibrations``` and ```capid``` below are only needed ```if (doInterCalib_)```
You can move their definition under that condition, e.g.:
```
      double time = rechit.time();
      double correctedenergy = factor_*rechit.energy();
      if (doInterCalib_) {
	 // do proper gain calibration reading the latest entries in the condDB
	 int capid = 0; // take some capid, gains are the same for all capid's
	 const CastorCalibrations& calibrations=conditions->getCastorCalibrations(rechit->id());
         correctedenergy *= calibrations.gain(capid);
      }

```

perrotta commented on RecoLocalCalo/Castor/src/RecHitCorrector.cc:
```diff
@@ -114,59 +114,47 @@ RecHitCorrector::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+	for (auto channel : channels) {	
+		if (channel.rawId() == detcell.rawId()) {
+			const CastorChannelStatus* mydigistatus=myqual->getValues(channel);
 			if (mydigistatus->getValue() == 2989) ok = false; // 2989 = BAD
```
You could even ```break``` here: no need to go through all remaining channels if ```ok``` is already ```false```

perrotta commented:
ping @hvanhaev

hvanhaev commented:
Thanks, for the reminder, I hope I can implement the suggestions in course of the next days...

On 15 Mar 2018, at 11:54, perrotta <notifications@github.com<mailto:notifications@github.com>> wrote:


ping @hvanhaev<https://github.com/hvanhaev>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/22424#issuecomment-373336795>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEgkFyWHJygxbX2YPzqz649wn20yXYd7ks5tekhogaJpZM4SaFZH>.

fabiocos commented:
@hvanhaev I understand that this PR is targeting studies for past years analysis, and it will have to be considered for 10_2_X at least. Could you please confirm or clarify?

hvanhaev commented:
Dear Fabio,

Yes indeed, if possible & allowed, it would be good to have:
- it implemented in a next CMSSW_10_X_Y so we can use it when CASTOR will participate in 2018 Heavy Ion running.
- it back ported to a release to be used with 2015 data (I assume 76X) and 2016 Heavy Ion data (I assume 80X).

This PR does not contain new features, but does have fixes for bugs that are important for CASTOR. It should also only affect CASTOR simulation and reconstruction (so no interference with other subsystems).

Cheers,
Hans


On 16 Mar 2018, at 14:14, Fabio Cossutti <notifications@github.com<mailto:notifications@github.com>> wrote:


@hvanhaev<https://github.com/hvanhaev> I understand that this PR is targeting studies for past years analysis, and it will have to be considered for 10_2_X at least. Could you please confirm or clarify?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/22424#issuecomment-373709266>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEgkFzDebhGxRXQc-FqD8JGteBmqsfw1ks5te7qpgaJpZM4SaFZH>.

perrotta commented:
Hans Van Haevermaet <notifications@github.com> ha scritto:

> Dear Fabio,
>
> Yes indeed, if possible & allowed, it would be good to have:
> - it implemented in a next CMSSW_10_X_Y so we can use it when CASTOR  
> will participate in 2018 Heavy Ion running.

As far I know the discussion whether to install CASTOR during TS2 is  
still ongoing: or was a decision taken by now?

By the way, all the CASTOR code that will eventually be allowed in the  
production sequences (not only these classes, I imagine) must be  
updated along the lines that we started to discuss in this PR

> - it back ported to a release to be used with 2015 data (I assume  
> 76X) and 2016 Heavy Ion data (I assume 80X).
>
> This PR does not contain new features, but does have fixes for bugs  
> that are important for CASTOR. It should also only affect CASTOR  
> simulation and reconstruction (so no interference with other  
> subsystems).
>
> Cheers,
> Hans
>
>
> On 16 Mar 2018, at 14:14, Fabio Cossutti  
> <notifications@github.com<mailto:notifications@github.com>> wrote:
>
>
> @hvanhaev<https://github.com/hvanhaev> I understand that this PR is  
> targeting studies for past years analysis, and it will have to be  
> considered for 10_2_X at least. Could you please confirm or clarify?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on  
> GitHub<https://github.com/cms-sw/cmssw/pull/22424#issuecomment-373709266>,  
> or mute the  
> thread<https://github.com/notifications/unsubscribe-auth/AEgkFzDebhGxRXQc-FqD8JGteBmqsfw1ks5te7qpgaJpZM4SaFZH>.
>
>
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/cms-sw/cmssw/pull/22424#issuecomment-373715071

hvanhaev commented:
As far I know the discussion whether to install CASTOR during TS2 is
still ongoing: or was a decision taken by now?

I don’t think it is indeed decided whether it will be installed during TS2 or TS3, but I do know that the aim of the CASTOR group is to participate in the HIN run this year, which is after TS3.
In any case, your comment on the previous PR implied that I first need to submit to master, and then request back ports...

By the way, all the CASTOR code that will eventually be allowed in the
production sequences (not only these classes, I imagine) must be
updated along the lines that we started to discuss in this PR

Basically all CASTOR code (except this RecHitCorrector, InvalidDataFilter and old cluster code) is already in production sequences since many years.
If I understand you correctly you like that we do a complete cross check of all our code (sim, digi, reconstruction, alca, …) to update according to new C++ standards?



On 16 Mar 2018, at 15:01, perrotta <notifications@github.com<mailto:notifications@github.com>> wrote:


Hans Van Haevermaet <notifications@github.com<mailto:notifications@github.com>> ha scritto:

> Dear Fabio,
>
> Yes indeed, if possible & allowed, it would be good to have:
> - it implemented in a next CMSSW_10_X_Y so we can use it when CASTOR
> will participate in 2018 Heavy Ion running.

As far I know the discussion whether to install CASTOR during TS2 is
still ongoing: or was a decision taken by now?

By the way, all the CASTOR code that will eventually be allowed in the
production sequences (not only these classes, I imagine) must be
updated along the lines that we started to discuss in this PR

> - it back ported to a release to be used with 2015 data (I assume
> 76X) and 2016 Heavy Ion data (I assume 80X).
>
> This PR does not contain new features, but does have fixes for bugs
> that are important for CASTOR. It should also only affect CASTOR
> simulation and reconstruction (so no interference with other
> subsystems).
>
> Cheers,
> Hans
>
>
> On 16 Mar 2018, at 14:14, Fabio Cossutti
> <notifications@github.com<mailto:notifications@github.com><mailto:notifications@github.com>> wrote:
>
>
> @hvanhaev<https://github.com/hvanhaev> I understand that this PR is
> targeting studies for past years analysis, and it will have to be
> considered for 10_2_X at least. Could you please confirm or clarify?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on
> GitHub<https://github.com/cms-sw/cmssw/pull/22424#issuecomment-373709266>,
> or mute the
> thread<https://github.com/notifications/unsubscribe-auth/AEgkFzDebhGxRXQc-FqD8JGteBmqsfw1ks5te7qpgaJpZM4SaFZH>.
>
>
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/cms-sw/cmssw/pull/22424#issuecomment-373715071




—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/22424#issuecomment-373722018>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEgkF8Z0t_PUY1ehOSxOjb8EhadyygZOks5te8W1gaJpZM4SaFZH>.

hvanhaev committed:
clean up code

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22424/4048

cmsbuild commented:
Pull request #22424 was updated. @perrotta, @cmsbuild, @civanch, @mdhildreth, @slava77 can you please check and sign again.

hvanhaev commented:
Hi Andrea, 
I committed your proposed changes to RecHitCorrector. Now it should be a stream module (?), I also removed the unused begin/endJob functions. Please let me know if you have further comments.
Cheers,
Hans

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 4cb8ca24993c06bf4dab95b4926241171735b4af
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/26993/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/af2458487e0e26a2bbb6eb8ac0f0c30c00993b6f
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/26993/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/26993/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/26993/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 78 differences found in the comparisons
 - DQMHistoTests: Total files compared: 29
 - DQMHistoTests: Total histograms compared: 2497880
 - DQMHistoTests: Total failures: 84
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2497620
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.83000000006 KiB( 22 files compared)
 - Checked 118 log files, 9 edm output root files, 29 DQM output files

perrotta commented:
+1
- Changes/fixes inside ```RecoLocalCalo/Castor/src/RecHitCorrector.cc``` are fine and comply with the requests received; that package does not enter currently in the standard reco sequence
- Changes are observed in the CastorRecHits and CastorTower reco outputs: they are due to the fixes applied here to the CASTOR Simulation packages, e..g. in TTbar13TeVPU2017 wf 10224.0:
![image](https://user-images.githubusercontent.com/4069749/37707159-6f8e01ce-2d02-11e8-9964-cd0d92acf17f.png)

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
@fabiocos : I think this can be merged in the master, if you don't see issues for it

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 4cb8ca24993c06bf4dab95b4926241171735b4af
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/27279/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/af2458487e0e26a2bbb6eb8ac0f0c30c00993b6f
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/27279/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/27279/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22424/27279/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 76 differences found in the comparisons
 - DQMHistoTests: Total files compared: 29
 - DQMHistoTests: Total histograms compared: 2498516
 - DQMHistoTests: Total failures: 83
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2498257
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.840000000033 KiB( 23 files compared)
 - Checked 119 log files, 9 edm output root files, 29 DQM output files

fabiocos commented:
@civanch @ianna CASTOR is presently not installed, it will be for the HI run, but it was kept anyway in the 2018 geometry configuration, even if TOTEM T1 and T2 were removed (see https://indico.cern.ch/event/714398/contributions/2938076/attachments/1618558/2573716/VI20180316sum.pdf ). I understand that this was done under the assumption that there is not effect to be expected on HF from possible back-scattering from it, right? So it is not useful but it does not harm either, is this the assumption?

@hvanhaev @slava77 @perrotta I see that also the 2017 workflows are affected by these changes, no era was used. From the initial comment I understand that the updates should be considered as bug fixes to be back-ported in 2015 and 2016 simulations, so it is ok that not only the 2018 behaviour is modified. Do you confirm?

civanch commented:
@fabiocos , both CASTOR and ZDS simulations are performed using shower libraries. This means that if a particle (with whatever energy) hit one of this calorimeters it is killed and set of hits are produced. So, they physically cannot backscatter to HF. In the #22145 this will not be true for Hyperons - they should first decay and only after SL will be activated, however, it is difficult to assume that a Hyperon will backscatter and path few meters instead of decay.

Concerning TOTEM: last year we missed the fact, that T1 detector was removed before 2017 run. In 2018 T2 was removed. The most effect come from T1 removal, because it is in front. Note, that HF shower library was created in a special simulation when primary particles hit HF alone, so HF SL is more close to the reality, when TOTEM T1 and T2 are removed from the geometry.

perrotta commented:
@fabiocos : yes, the changes in 2017 wfs (also observed in https://github.com/cms-sw/cmssw/pull/22424#issuecomment-374907797) come from the fixes for the previously overestimated noise in the CASTOR simulation, and they are supposed to be applied whenever the CASTOR simulation is run

fabiocos commented:
@hvanhaev @civanch @perrotta thanks for the clarifications, I will integrate this PR with other SIM-related developments after 10_2_0_pre1 is built

fabiocos commented:
+1
